internal/xds/translator/authorization.go

```diff
@@ -262,6 +278,243 @@ func (*rbac) patchRoute(route *routev3.Route, irRoute *ir.HTTPRoute) error {
 	return nil
 }
 
+func buildIPPredicate(clientCIDRs []*ir.CIDRMatch) (*matcherv3.Matcher_MatcherList_Predicate_SinglePredicate_, error) {
+	var (
+		sourceIPInput *anypb.Any
+		ipMatcher     *anypb.Any
+		err           error
+	)
+
+	// Build the IPMatcher based on the client CIDRs.
+	ipRangeMatcher := &ipmatcherv3.Ip{
+		StatPrefix: "client_ip",
+	}
+
+	for _, cidr := range clientCIDRs {
+		ipRangeMatcher.CidrRanges = append(ipRangeMatcher.CidrRanges, &configv3.CidrRange{
+			AddressPrefix: cidr.IP,
+			PrefixLen: &wrapperspb.UInt32Value{
+				Value: cidr.MaskLen,
+			},
+		})
+	}
+
+	if ipMatcher, err = anypb.New(ipRangeMatcher); err != nil {
+		return nil, err
+	}
+
+	if sourceIPInput, err = anypb.New(&networkinput.SourceIPInput{}); err != nil {
+		return nil, err
+	}
+
+	return &matcherv3.Matcher_MatcherList_Predicate_SinglePredicate_{
+		SinglePredicate: &matcherv3.Matcher_MatcherList_Predicate_SinglePredicate{
+			Input: &cncfv3.TypedExtensionConfig{
+				Name:        "client_ip",
+				TypedConfig: sourceIPInput,
+			},
+			Matcher: &matcherv3.Matcher_MatcherList_Predicate_SinglePredicate_CustomMatch{
+				CustomMatch: &cncfv3.TypedExtensionConfig{
+					Name:        "ip_matcher",
+					TypedConfig: ipMatcher,
+				},
+			},
+		},
+	}, nil
+}
+
+func buildJWTPredicate(jwt egv1a1.JWTPrincipal) ([]*matcherv3.Matcher_MatcherList_Predicate, error) {
+	jwtPredicate := []*matcherv3.Matcher_MatcherList_Predicate{}
+
+	// Build the scope matchers.
+	// Multiple scopes are ANDed together.
+	for _, scope := range jwt.Scopes {
+		var (
+			inputPb   *anypb.Any
+			matcherPb *anypb.Any
+			err       error
+		)
+
+		input := &networkinput.DynamicMetadataInput{
+			Filter: "envoy.filters.http.jwt_authn",
+			Path: []*networkinput.DynamicMetadataInput_PathSegment{
+				{
+					Segment: &networkinput.DynamicMetadataInput_PathSegment_Key{
+						Key: jwt.Provider, // The issuer is used as the `payload_in_metadata` in the JWT Authn filter.
+					},
+				},
+				{
+					Segment: &networkinput.DynamicMetadataInput_PathSegment_Key{
+						Key: "scope",
+					},
+				},
+			},
+		}
+
+		// The scope is a space-delimited string, so we use a regex matcher to match the scope.
+		pattern := fmt.Sprintf(`(^|.*\s)%s(\s.*|$)`, scope)
```

Comment:
@denniskniep had presented a nice suggestion earlier of using https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/jwt_authn/v3/config.proto#extensions-filters-http-jwt-authn-v3-jwtprovider-normalizepayload, that may make this faster